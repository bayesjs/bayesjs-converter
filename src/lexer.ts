import { Token } from './types';

const regexWhitespace = /\s/;
const regexNumber = /[0-9]/;
const regexIdentifier = /[a-z0-9_]/i;

export default (source: string): Array<Token> => {
  const tokens: Array<Token> = [];

  source = source.replace(/\\303\\243/g, 'Ã£');

  let currentIndex = 0;

  while (currentIndex < source.length) {
    let char = source[currentIndex];

    if (char === '%') {
      while (char !== '\n') {
        char = source[++currentIndex];
      }

      continue;
    }

    if (char === '=') {
      tokens.push({ type: 'equal' });
      currentIndex++;
      continue;
    }

    if (char === ';') {
      tokens.push({ type: 'semicolon' });
      currentIndex++;
      continue;
    }

    if (char === '|') {
      tokens.push({ type: 'pipe' });
      currentIndex++;
      continue;
    }

    if (char === '(') {
      tokens.push({ type: 'openparen' });
      currentIndex++;
      continue;
    }

    if (char === ')') {
      tokens.push({ type: 'closeparen' });
      currentIndex++;
      continue;
    }

    if (char === '{') {
      tokens.push({ type: 'openbracket' });
      currentIndex++;
      continue;
    }

    if (char === '}') {
      tokens.push({ type: 'closebracket' });
      currentIndex++;
      continue;
    }

    if (regexWhitespace.test(char)) {
      currentIndex++;
      continue;
    }

    if (char === '"') {
      let value = '';

      char = source[++currentIndex];

      while (char !== '"') {
        value += char;
        char = source[++currentIndex];
      }

      tokens.push({ type: 'string', value });
      currentIndex++;
      continue;
    }

    if (regexNumber.test(char)) {
      let value = '';
      let foundDecimalSeparator = false;

      while (
        regexNumber.test(char) ||
        (char === '.' && foundDecimalSeparator === false)
      ) {
        if (char === '.') {
          foundDecimalSeparator = true;
        }

        value += char;
        char = source[++currentIndex];
      }

      tokens.push({ type: 'number', value: parseFloat(value) });
      continue;
    }

    if (regexIdentifier.test(char)) {
      let value = '';

      while (regexIdentifier.test(char)) {
        value += char;
        char = source[++currentIndex];
      }

      switch (value) {
        case 'net':
          tokens.push({ type: 'net' });
          break;
        case 'node':
          tokens.push({ type: 'node' });
          break;
        case 'potential':
          tokens.push({ type: 'potential' });
          break;
        default:
          tokens.push({ type: 'identifier', value });
          break;
      }

      continue;
    }

    throw new Error(`Unexpected character: ${char}`);
  }

  return tokens;
};
