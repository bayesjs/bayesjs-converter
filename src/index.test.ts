import * as fs from 'fs';
import * as path from 'path';
import lexer from './lexer';
import parser from './parser';

it('converts network correctly', () => {
  const source = readNetwork('chuva');
  const tokens = lexer(source);
  const ast = parser(tokens);

  expect(tokens).toMatchSnapshot();
  expect(ast).toMatchSnapshot();
});

const readNetwork = (networkName: string) => {
  const networkPath = path.join(__dirname, `../networks/${networkName}.net`);
  const networkSource = fs.readFileSync(networkPath).toString();

  return networkSource;
};
