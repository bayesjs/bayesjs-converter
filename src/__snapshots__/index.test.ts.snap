// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`converts network correctly 1`] = `
Array [
  Object {
    "type": "net",
  },
  Object {
    "type": "openbracket",
  },
  Object {
    "type": "identifier",
    "value": "node_size",
  },
  Object {
    "type": "equal",
  },
  Object {
    "type": "openparen",
  },
  Object {
    "type": "number",
    "value": 80,
  },
  Object {
    "type": "number",
    "value": 40,
  },
  Object {
    "type": "closeparen",
  },
  Object {
    "type": "semicolon",
  },
  Object {
    "type": "identifier",
    "value": "HR_DBN_Num_Slices",
  },
  Object {
    "type": "equal",
  },
  Object {
    "type": "string",
    "value": "5",
  },
  Object {
    "type": "semicolon",
  },
  Object {
    "type": "identifier",
    "value": "HR_Desc",
  },
  Object {
    "type": "equal",
  },
  Object {
    "type": "string",
    "value": "",
  },
  Object {
    "type": "semicolon",
  },
  Object {
    "type": "identifier",
    "value": "HR_OOBN_CheckConsistency",
  },
  Object {
    "type": "equal",
  },
  Object {
    "type": "string",
    "value": "1",
  },
  Object {
    "type": "semicolon",
  },
  Object {
    "type": "identifier",
    "value": "HR_OOBN_FitExpand",
  },
  Object {
    "type": "equal",
  },
  Object {
    "type": "string",
    "value": "1",
  },
  Object {
    "type": "semicolon",
  },
  Object {
    "type": "identifier",
    "value": "HR_Zoom_ZoomFactor",
  },
  Object {
    "type": "equal",
  },
  Object {
    "type": "string",
    "value": "100",
  },
  Object {
    "type": "semicolon",
  },
  Object {
    "type": "identifier",
    "value": "HR_Groups_GroupColors",
  },
  Object {
    "type": "equal",
  },
  Object {
    "type": "string",
    "value": "",
  },
  Object {
    "type": "semicolon",
  },
  Object {
    "type": "identifier",
    "value": "HR_Groups_GroupNames",
  },
  Object {
    "type": "equal",
  },
  Object {
    "type": "string",
    "value": "",
  },
  Object {
    "type": "semicolon",
  },
  Object {
    "type": "identifier",
    "value": "HR_Groups_UserGroupsNo",
  },
  Object {
    "type": "equal",
  },
  Object {
    "type": "string",
    "value": "0",
  },
  Object {
    "type": "semicolon",
  },
  Object {
    "type": "identifier",
    "value": "HR_Utility_Link",
  },
  Object {
    "type": "equal",
  },
  Object {
    "type": "string",
    "value": "63",
  },
  Object {
    "type": "semicolon",
  },
  Object {
    "type": "identifier",
    "value": "HR_Probabilistic_Link",
  },
  Object {
    "type": "equal",
  },
  Object {
    "type": "string",
    "value": "63",
  },
  Object {
    "type": "semicolon",
  },
  Object {
    "type": "identifier",
    "value": "HR_Functional_Link",
  },
  Object {
    "type": "equal",
  },
  Object {
    "type": "string",
    "value": "63",
  },
  Object {
    "type": "semicolon",
  },
  Object {
    "type": "identifier",
    "value": "HR_Binding_Link",
  },
  Object {
    "type": "equal",
  },
  Object {
    "type": "string",
    "value": "63",
  },
  Object {
    "type": "semicolon",
  },
  Object {
    "type": "identifier",
    "value": "HR_Information_Arc",
  },
  Object {
    "type": "equal",
  },
  Object {
    "type": "string",
    "value": "63",
  },
  Object {
    "type": "semicolon",
  },
  Object {
    "type": "identifier",
    "value": "HR_Color_Interface",
  },
  Object {
    "type": "equal",
  },
  Object {
    "type": "string",
    "value": "21",
  },
  Object {
    "type": "semicolon",
  },
  Object {
    "type": "identifier",
    "value": "HR_Color_Instance",
  },
  Object {
    "type": "equal",
  },
  Object {
    "type": "string",
    "value": "0",
  },
  Object {
    "type": "semicolon",
  },
  Object {
    "type": "identifier",
    "value": "HR_Color_DiscreteFunction",
  },
  Object {
    "type": "equal",
  },
  Object {
    "type": "string",
    "value": "5",
  },
  Object {
    "type": "semicolon",
  },
  Object {
    "type": "identifier",
    "value": "HR_Color_Function",
  },
  Object {
    "type": "equal",
  },
  Object {
    "type": "string",
    "value": "4",
  },
  Object {
    "type": "semicolon",
  },
  Object {
    "type": "identifier",
    "value": "HR_Color_Utility",
  },
  Object {
    "type": "equal",
  },
  Object {
    "type": "string",
    "value": "36",
  },
  Object {
    "type": "semicolon",
  },
  Object {
    "type": "identifier",
    "value": "HR_Color_Decision",
  },
  Object {
    "type": "equal",
  },
  Object {
    "type": "string",
    "value": "17",
  },
  Object {
    "type": "semicolon",
  },
  Object {
    "type": "identifier",
    "value": "HR_Color_ContinuousChance",
  },
  Object {
    "type": "equal",
  },
  Object {
    "type": "string",
    "value": "48",
  },
  Object {
    "type": "semicolon",
  },
  Object {
    "type": "identifier",
    "value": "HR_Color_DiscreteChance",
  },
  Object {
    "type": "equal",
  },
  Object {
    "type": "string",
    "value": "16",
  },
  Object {
    "type": "semicolon",
  },
  Object {
    "type": "identifier",
    "value": "HR_Monitor_InitSD",
  },
  Object {
    "type": "equal",
  },
  Object {
    "type": "string",
    "value": "2",
  },
  Object {
    "type": "semicolon",
  },
  Object {
    "type": "identifier",
    "value": "HR_Monitor_Utility_Arrange",
  },
  Object {
    "type": "equal",
  },
  Object {
    "type": "string",
    "value": "0",
  },
  Object {
    "type": "semicolon",
  },
  Object {
    "type": "identifier",
    "value": "HR_Monitor_Mean_Variance",
  },
  Object {
    "type": "equal",
  },
  Object {
    "type": "string",
    "value": "1",
  },
  Object {
    "type": "semicolon",
  },
  Object {
    "type": "identifier",
    "value": "HR_Monitor_Chance_Utility",
  },
  Object {
    "type": "equal",
  },
  Object {
    "type": "string",
    "value": "1",
  },
  Object {
    "type": "semicolon",
  },
  Object {
    "type": "identifier",
    "value": "HR_Monitor_InitStates",
  },
  Object {
    "type": "equal",
  },
  Object {
    "type": "string",
    "value": "5",
  },
  Object {
    "type": "semicolon",
  },
  Object {
    "type": "identifier",
    "value": "HR_Monitor_OpenGraph",
  },
  Object {
    "type": "equal",
  },
  Object {
    "type": "string",
    "value": "0",
  },
  Object {
    "type": "semicolon",
  },
  Object {
    "type": "identifier",
    "value": "HR_Monitor_GraphPrecision",
  },
  Object {
    "type": "equal",
  },
  Object {
    "type": "string",
    "value": "100",
  },
  Object {
    "type": "semicolon",
  },
  Object {
    "type": "identifier",
    "value": "HR_Monitor_AutoUpdGraph",
  },
  Object {
    "type": "equal",
  },
  Object {
    "type": "string",
    "value": "0",
  },
  Object {
    "type": "semicolon",
  },
  Object {
    "type": "identifier",
    "value": "HR_Compile_MaxMinimalSeparators",
  },
  Object {
    "type": "equal",
  },
  Object {
    "type": "string",
    "value": "100000",
  },
  Object {
    "type": "semicolon",
  },
  Object {
    "type": "identifier",
    "value": "HR_Compile_ApproxEpsilon",
  },
  Object {
    "type": "equal",
  },
  Object {
    "type": "string",
    "value": "1.0E-5",
  },
  Object {
    "type": "semicolon",
  },
  Object {
    "type": "identifier",
    "value": "HR_Compile_Approximate",
  },
  Object {
    "type": "equal",
  },
  Object {
    "type": "string",
    "value": "0",
  },
  Object {
    "type": "semicolon",
  },
  Object {
    "type": "identifier",
    "value": "HR_Compile_SaveToMemory",
  },
  Object {
    "type": "equal",
  },
  Object {
    "type": "string",
    "value": "0",
  },
  Object {
    "type": "semicolon",
  },
  Object {
    "type": "identifier",
    "value": "HR_Compile_Compress",
  },
  Object {
    "type": "equal",
  },
  Object {
    "type": "string",
    "value": "0",
  },
  Object {
    "type": "semicolon",
  },
  Object {
    "type": "identifier",
    "value": "HR_Compile_TriangMethod",
  },
  Object {
    "type": "equal",
  },
  Object {
    "type": "string",
    "value": "0",
  },
  Object {
    "type": "semicolon",
  },
  Object {
    "type": "identifier",
    "value": "HR_Propagate_AutoNormal",
  },
  Object {
    "type": "equal",
  },
  Object {
    "type": "string",
    "value": "1",
  },
  Object {
    "type": "semicolon",
  },
  Object {
    "type": "identifier",
    "value": "HR_Propagate_AutoSum",
  },
  Object {
    "type": "equal",
  },
  Object {
    "type": "string",
    "value": "1",
  },
  Object {
    "type": "semicolon",
  },
  Object {
    "type": "identifier",
    "value": "HR_Propagate_Auto",
  },
  Object {
    "type": "equal",
  },
  Object {
    "type": "string",
    "value": "1",
  },
  Object {
    "type": "semicolon",
  },
  Object {
    "type": "identifier",
    "value": "HR_Font_Italic",
  },
  Object {
    "type": "equal",
  },
  Object {
    "type": "string",
    "value": "0",
  },
  Object {
    "type": "semicolon",
  },
  Object {
    "type": "identifier",
    "value": "HR_Font_Weight",
  },
  Object {
    "type": "equal",
  },
  Object {
    "type": "string",
    "value": "400",
  },
  Object {
    "type": "semicolon",
  },
  Object {
    "type": "identifier",
    "value": "HR_Font_Size",
  },
  Object {
    "type": "equal",
  },
  Object {
    "type": "string",
    "value": "-12",
  },
  Object {
    "type": "semicolon",
  },
  Object {
    "type": "identifier",
    "value": "HR_Font_Name",
  },
  Object {
    "type": "equal",
  },
  Object {
    "type": "string",
    "value": "javax.swing.plaf.FontUIResource[family=Tahoma,name=Tahoma,style=plain,size=11]",
  },
  Object {
    "type": "semicolon",
  },
  Object {
    "type": "identifier",
    "value": "HR_Grid_GridShow",
  },
  Object {
    "type": "equal",
  },
  Object {
    "type": "string",
    "value": "0",
  },
  Object {
    "type": "semicolon",
  },
  Object {
    "type": "identifier",
    "value": "HR_Grid_GridSnap",
  },
  Object {
    "type": "equal",
  },
  Object {
    "type": "string",
    "value": "0",
  },
  Object {
    "type": "semicolon",
  },
  Object {
    "type": "identifier",
    "value": "HR_Grid_Y",
  },
  Object {
    "type": "equal",
  },
  Object {
    "type": "string",
    "value": "40",
  },
  Object {
    "type": "semicolon",
  },
  Object {
    "type": "identifier",
    "value": "HR_Grid_X",
  },
  Object {
    "type": "equal",
  },
  Object {
    "type": "string",
    "value": "40",
  },
  Object {
    "type": "semicolon",
  },
  Object {
    "type": "closebracket",
  },
  Object {
    "type": "node",
  },
  Object {
    "type": "identifier",
    "value": "GRAMA_MOLHADA",
  },
  Object {
    "type": "openbracket",
  },
  Object {
    "type": "identifier",
    "value": "label",
  },
  Object {
    "type": "equal",
  },
  Object {
    "type": "string",
    "value": "",
  },
  Object {
    "type": "semicolon",
  },
  Object {
    "type": "identifier",
    "value": "position",
  },
  Object {
    "type": "equal",
  },
  Object {
    "type": "openparen",
  },
  Object {
    "type": "number",
    "value": 283,
  },
  Object {
    "type": "number",
    "value": 82,
  },
  Object {
    "type": "closeparen",
  },
  Object {
    "type": "semicolon",
  },
  Object {
    "type": "identifier",
    "value": "states",
  },
  Object {
    "type": "equal",
  },
  Object {
    "type": "openparen",
  },
  Object {
    "type": "string",
    "value": "Sim",
  },
  Object {
    "type": "string",
    "value": "Não",
  },
  Object {
    "type": "closeparen",
  },
  Object {
    "type": "semicolon",
  },
  Object {
    "type": "identifier",
    "value": "HR_NodeSize",
  },
  Object {
    "type": "equal",
  },
  Object {
    "type": "string",
    "value": "(141, 59)",
  },
  Object {
    "type": "semicolon",
  },
  Object {
    "type": "identifier",
    "value": "HR_Group",
  },
  Object {
    "type": "equal",
  },
  Object {
    "type": "string",
    "value": "0",
  },
  Object {
    "type": "semicolon",
  },
  Object {
    "type": "identifier",
    "value": "HR_Desc",
  },
  Object {
    "type": "equal",
  },
  Object {
    "type": "string",
    "value": "",
  },
  Object {
    "type": "semicolon",
  },
  Object {
    "type": "identifier",
    "value": "HR_State_1",
  },
  Object {
    "type": "equal",
  },
  Object {
    "type": "string",
    "value": "",
  },
  Object {
    "type": "semicolon",
  },
  Object {
    "type": "identifier",
    "value": "HR_State_0",
  },
  Object {
    "type": "equal",
  },
  Object {
    "type": "string",
    "value": "",
  },
  Object {
    "type": "semicolon",
  },
  Object {
    "type": "closebracket",
  },
  Object {
    "type": "node",
  },
  Object {
    "type": "identifier",
    "value": "REGADOR",
  },
  Object {
    "type": "openbracket",
  },
  Object {
    "type": "identifier",
    "value": "label",
  },
  Object {
    "type": "equal",
  },
  Object {
    "type": "string",
    "value": "",
  },
  Object {
    "type": "semicolon",
  },
  Object {
    "type": "identifier",
    "value": "position",
  },
  Object {
    "type": "equal",
  },
  Object {
    "type": "openparen",
  },
  Object {
    "type": "number",
    "value": 147,
  },
  Object {
    "type": "number",
    "value": 185,
  },
  Object {
    "type": "closeparen",
  },
  Object {
    "type": "semicolon",
  },
  Object {
    "type": "identifier",
    "value": "states",
  },
  Object {
    "type": "equal",
  },
  Object {
    "type": "openparen",
  },
  Object {
    "type": "string",
    "value": "Sim",
  },
  Object {
    "type": "string",
    "value": "Não",
  },
  Object {
    "type": "closeparen",
  },
  Object {
    "type": "semicolon",
  },
  Object {
    "type": "identifier",
    "value": "HR_LinkGroup",
  },
  Object {
    "type": "equal",
  },
  Object {
    "type": "string",
    "value": "[GRAMA_MOLHADA:0]",
  },
  Object {
    "type": "semicolon",
  },
  Object {
    "type": "identifier",
    "value": "HR_LinkMode",
  },
  Object {
    "type": "equal",
  },
  Object {
    "type": "string",
    "value": "[GRAMA_MOLHADA:0]",
  },
  Object {
    "type": "semicolon",
  },
  Object {
    "type": "identifier",
    "value": "HR_Group",
  },
  Object {
    "type": "equal",
  },
  Object {
    "type": "string",
    "value": "0",
  },
  Object {
    "type": "semicolon",
  },
  Object {
    "type": "identifier",
    "value": "HR_Desc",
  },
  Object {
    "type": "equal",
  },
  Object {
    "type": "string",
    "value": "",
  },
  Object {
    "type": "semicolon",
  },
  Object {
    "type": "identifier",
    "value": "HR_State_1",
  },
  Object {
    "type": "equal",
  },
  Object {
    "type": "string",
    "value": "",
  },
  Object {
    "type": "semicolon",
  },
  Object {
    "type": "identifier",
    "value": "HR_State_0",
  },
  Object {
    "type": "equal",
  },
  Object {
    "type": "string",
    "value": "",
  },
  Object {
    "type": "semicolon",
  },
  Object {
    "type": "closebracket",
  },
  Object {
    "type": "node",
  },
  Object {
    "type": "identifier",
    "value": "CHUVA",
  },
  Object {
    "type": "openbracket",
  },
  Object {
    "type": "identifier",
    "value": "label",
  },
  Object {
    "type": "equal",
  },
  Object {
    "type": "string",
    "value": "",
  },
  Object {
    "type": "semicolon",
  },
  Object {
    "type": "identifier",
    "value": "position",
  },
  Object {
    "type": "equal",
  },
  Object {
    "type": "openparen",
  },
  Object {
    "type": "number",
    "value": 397,
  },
  Object {
    "type": "number",
    "value": 190,
  },
  Object {
    "type": "closeparen",
  },
  Object {
    "type": "semicolon",
  },
  Object {
    "type": "identifier",
    "value": "states",
  },
  Object {
    "type": "equal",
  },
  Object {
    "type": "openparen",
  },
  Object {
    "type": "string",
    "value": "Sim",
  },
  Object {
    "type": "string",
    "value": "Não",
  },
  Object {
    "type": "closeparen",
  },
  Object {
    "type": "semicolon",
  },
  Object {
    "type": "identifier",
    "value": "HR_LinkGroup",
  },
  Object {
    "type": "equal",
  },
  Object {
    "type": "string",
    "value": "[REGADOR:0][GRAMA_MOLHADA:0]",
  },
  Object {
    "type": "semicolon",
  },
  Object {
    "type": "identifier",
    "value": "HR_LinkMode",
  },
  Object {
    "type": "equal",
  },
  Object {
    "type": "string",
    "value": "[REGADOR:0][GRAMA_MOLHADA:0]",
  },
  Object {
    "type": "semicolon",
  },
  Object {
    "type": "identifier",
    "value": "HR_Group",
  },
  Object {
    "type": "equal",
  },
  Object {
    "type": "string",
    "value": "0",
  },
  Object {
    "type": "semicolon",
  },
  Object {
    "type": "identifier",
    "value": "HR_Desc",
  },
  Object {
    "type": "equal",
  },
  Object {
    "type": "string",
    "value": "",
  },
  Object {
    "type": "semicolon",
  },
  Object {
    "type": "identifier",
    "value": "HR_State_1",
  },
  Object {
    "type": "equal",
  },
  Object {
    "type": "string",
    "value": "",
  },
  Object {
    "type": "semicolon",
  },
  Object {
    "type": "identifier",
    "value": "HR_State_0",
  },
  Object {
    "type": "equal",
  },
  Object {
    "type": "string",
    "value": "",
  },
  Object {
    "type": "semicolon",
  },
  Object {
    "type": "closebracket",
  },
  Object {
    "type": "potential",
  },
  Object {
    "type": "openparen",
  },
  Object {
    "type": "identifier",
    "value": "GRAMA_MOLHADA",
  },
  Object {
    "type": "pipe",
  },
  Object {
    "type": "identifier",
    "value": "REGADOR",
  },
  Object {
    "type": "identifier",
    "value": "CHUVA",
  },
  Object {
    "type": "closeparen",
  },
  Object {
    "type": "openbracket",
  },
  Object {
    "type": "identifier",
    "value": "data",
  },
  Object {
    "type": "equal",
  },
  Object {
    "type": "openparen",
  },
  Object {
    "type": "openparen",
  },
  Object {
    "type": "openparen",
  },
  Object {
    "type": "number",
    "value": 0.99,
  },
  Object {
    "type": "number",
    "value": 0.01,
  },
  Object {
    "type": "closeparen",
  },
  Object {
    "type": "openparen",
  },
  Object {
    "type": "number",
    "value": 0.9,
  },
  Object {
    "type": "number",
    "value": 0.1,
  },
  Object {
    "type": "closeparen",
  },
  Object {
    "type": "closeparen",
  },
  Object {
    "type": "openparen",
  },
  Object {
    "type": "openparen",
  },
  Object {
    "type": "number",
    "value": 0.8,
  },
  Object {
    "type": "number",
    "value": 0.2,
  },
  Object {
    "type": "closeparen",
  },
  Object {
    "type": "openparen",
  },
  Object {
    "type": "number",
    "value": 0,
  },
  Object {
    "type": "number",
    "value": 1,
  },
  Object {
    "type": "closeparen",
  },
  Object {
    "type": "closeparen",
  },
  Object {
    "type": "closeparen",
  },
  Object {
    "type": "semicolon",
  },
  Object {
    "type": "closebracket",
  },
  Object {
    "type": "potential",
  },
  Object {
    "type": "openparen",
  },
  Object {
    "type": "identifier",
    "value": "REGADOR",
  },
  Object {
    "type": "pipe",
  },
  Object {
    "type": "identifier",
    "value": "CHUVA",
  },
  Object {
    "type": "closeparen",
  },
  Object {
    "type": "openbracket",
  },
  Object {
    "type": "identifier",
    "value": "data",
  },
  Object {
    "type": "equal",
  },
  Object {
    "type": "openparen",
  },
  Object {
    "type": "openparen",
  },
  Object {
    "type": "number",
    "value": 0.01,
  },
  Object {
    "type": "number",
    "value": 0.99,
  },
  Object {
    "type": "closeparen",
  },
  Object {
    "type": "openparen",
  },
  Object {
    "type": "number",
    "value": 0.4,
  },
  Object {
    "type": "number",
    "value": 0.6,
  },
  Object {
    "type": "closeparen",
  },
  Object {
    "type": "closeparen",
  },
  Object {
    "type": "semicolon",
  },
  Object {
    "type": "closebracket",
  },
  Object {
    "type": "potential",
  },
  Object {
    "type": "openparen",
  },
  Object {
    "type": "identifier",
    "value": "CHUVA",
  },
  Object {
    "type": "closeparen",
  },
  Object {
    "type": "openbracket",
  },
  Object {
    "type": "identifier",
    "value": "data",
  },
  Object {
    "type": "equal",
  },
  Object {
    "type": "openparen",
  },
  Object {
    "type": "number",
    "value": 0.2,
  },
  Object {
    "type": "number",
    "value": 0.8,
  },
  Object {
    "type": "closeparen",
  },
  Object {
    "type": "semicolon",
  },
  Object {
    "type": "closebracket",
  },
]
`;

exports[`converts network correctly 2`] = `
Object {
  "items": Array [
    Object {
      "properties": Array [
        Object {
          "name": "node_size",
          "type": "property",
          "value": Array [
            80,
            40,
          ],
        },
        Object {
          "name": "HR_DBN_Num_Slices",
          "type": "property",
          "value": "5",
        },
        Object {
          "name": "HR_Desc",
          "type": "property",
          "value": "",
        },
        Object {
          "name": "HR_OOBN_CheckConsistency",
          "type": "property",
          "value": "1",
        },
        Object {
          "name": "HR_OOBN_FitExpand",
          "type": "property",
          "value": "1",
        },
        Object {
          "name": "HR_Zoom_ZoomFactor",
          "type": "property",
          "value": "100",
        },
        Object {
          "name": "HR_Groups_GroupColors",
          "type": "property",
          "value": "",
        },
        Object {
          "name": "HR_Groups_GroupNames",
          "type": "property",
          "value": "",
        },
        Object {
          "name": "HR_Groups_UserGroupsNo",
          "type": "property",
          "value": "0",
        },
        Object {
          "name": "HR_Utility_Link",
          "type": "property",
          "value": "63",
        },
        Object {
          "name": "HR_Probabilistic_Link",
          "type": "property",
          "value": "63",
        },
        Object {
          "name": "HR_Functional_Link",
          "type": "property",
          "value": "63",
        },
        Object {
          "name": "HR_Binding_Link",
          "type": "property",
          "value": "63",
        },
        Object {
          "name": "HR_Information_Arc",
          "type": "property",
          "value": "63",
        },
        Object {
          "name": "HR_Color_Interface",
          "type": "property",
          "value": "21",
        },
        Object {
          "name": "HR_Color_Instance",
          "type": "property",
          "value": "0",
        },
        Object {
          "name": "HR_Color_DiscreteFunction",
          "type": "property",
          "value": "5",
        },
        Object {
          "name": "HR_Color_Function",
          "type": "property",
          "value": "4",
        },
        Object {
          "name": "HR_Color_Utility",
          "type": "property",
          "value": "36",
        },
        Object {
          "name": "HR_Color_Decision",
          "type": "property",
          "value": "17",
        },
        Object {
          "name": "HR_Color_ContinuousChance",
          "type": "property",
          "value": "48",
        },
        Object {
          "name": "HR_Color_DiscreteChance",
          "type": "property",
          "value": "16",
        },
        Object {
          "name": "HR_Monitor_InitSD",
          "type": "property",
          "value": "2",
        },
        Object {
          "name": "HR_Monitor_Utility_Arrange",
          "type": "property",
          "value": "0",
        },
        Object {
          "name": "HR_Monitor_Mean_Variance",
          "type": "property",
          "value": "1",
        },
        Object {
          "name": "HR_Monitor_Chance_Utility",
          "type": "property",
          "value": "1",
        },
        Object {
          "name": "HR_Monitor_InitStates",
          "type": "property",
          "value": "5",
        },
        Object {
          "name": "HR_Monitor_OpenGraph",
          "type": "property",
          "value": "0",
        },
        Object {
          "name": "HR_Monitor_GraphPrecision",
          "type": "property",
          "value": "100",
        },
        Object {
          "name": "HR_Monitor_AutoUpdGraph",
          "type": "property",
          "value": "0",
        },
        Object {
          "name": "HR_Compile_MaxMinimalSeparators",
          "type": "property",
          "value": "100000",
        },
        Object {
          "name": "HR_Compile_ApproxEpsilon",
          "type": "property",
          "value": "1.0E-5",
        },
        Object {
          "name": "HR_Compile_Approximate",
          "type": "property",
          "value": "0",
        },
        Object {
          "name": "HR_Compile_SaveToMemory",
          "type": "property",
          "value": "0",
        },
        Object {
          "name": "HR_Compile_Compress",
          "type": "property",
          "value": "0",
        },
        Object {
          "name": "HR_Compile_TriangMethod",
          "type": "property",
          "value": "0",
        },
        Object {
          "name": "HR_Propagate_AutoNormal",
          "type": "property",
          "value": "1",
        },
        Object {
          "name": "HR_Propagate_AutoSum",
          "type": "property",
          "value": "1",
        },
        Object {
          "name": "HR_Propagate_Auto",
          "type": "property",
          "value": "1",
        },
        Object {
          "name": "HR_Font_Italic",
          "type": "property",
          "value": "0",
        },
        Object {
          "name": "HR_Font_Weight",
          "type": "property",
          "value": "400",
        },
        Object {
          "name": "HR_Font_Size",
          "type": "property",
          "value": "-12",
        },
        Object {
          "name": "HR_Font_Name",
          "type": "property",
          "value": "javax.swing.plaf.FontUIResource[family=Tahoma,name=Tahoma,style=plain,size=11]",
        },
        Object {
          "name": "HR_Grid_GridShow",
          "type": "property",
          "value": "0",
        },
        Object {
          "name": "HR_Grid_GridSnap",
          "type": "property",
          "value": "0",
        },
        Object {
          "name": "HR_Grid_Y",
          "type": "property",
          "value": "40",
        },
        Object {
          "name": "HR_Grid_X",
          "type": "property",
          "value": "40",
        },
      ],
      "type": "net",
    },
    Object {
      "name": "GRAMA_MOLHADA",
      "properties": Array [
        Object {
          "name": "label",
          "type": "property",
          "value": "",
        },
        Object {
          "name": "position",
          "type": "property",
          "value": Array [
            283,
            82,
          ],
        },
        Object {
          "name": "states",
          "type": "property",
          "value": Array [
            "Sim",
            "Não",
          ],
        },
        Object {
          "name": "HR_NodeSize",
          "type": "property",
          "value": "(141, 59)",
        },
        Object {
          "name": "HR_Group",
          "type": "property",
          "value": "0",
        },
        Object {
          "name": "HR_Desc",
          "type": "property",
          "value": "",
        },
        Object {
          "name": "HR_State_1",
          "type": "property",
          "value": "",
        },
        Object {
          "name": "HR_State_0",
          "type": "property",
          "value": "",
        },
      ],
      "type": "node",
    },
    Object {
      "name": "REGADOR",
      "properties": Array [
        Object {
          "name": "label",
          "type": "property",
          "value": "",
        },
        Object {
          "name": "position",
          "type": "property",
          "value": Array [
            147,
            185,
          ],
        },
        Object {
          "name": "states",
          "type": "property",
          "value": Array [
            "Sim",
            "Não",
          ],
        },
        Object {
          "name": "HR_LinkGroup",
          "type": "property",
          "value": "[GRAMA_MOLHADA:0]",
        },
        Object {
          "name": "HR_LinkMode",
          "type": "property",
          "value": "[GRAMA_MOLHADA:0]",
        },
        Object {
          "name": "HR_Group",
          "type": "property",
          "value": "0",
        },
        Object {
          "name": "HR_Desc",
          "type": "property",
          "value": "",
        },
        Object {
          "name": "HR_State_1",
          "type": "property",
          "value": "",
        },
        Object {
          "name": "HR_State_0",
          "type": "property",
          "value": "",
        },
      ],
      "type": "node",
    },
    Object {
      "name": "CHUVA",
      "properties": Array [
        Object {
          "name": "label",
          "type": "property",
          "value": "",
        },
        Object {
          "name": "position",
          "type": "property",
          "value": Array [
            397,
            190,
          ],
        },
        Object {
          "name": "states",
          "type": "property",
          "value": Array [
            "Sim",
            "Não",
          ],
        },
        Object {
          "name": "HR_LinkGroup",
          "type": "property",
          "value": "[REGADOR:0][GRAMA_MOLHADA:0]",
        },
        Object {
          "name": "HR_LinkMode",
          "type": "property",
          "value": "[REGADOR:0][GRAMA_MOLHADA:0]",
        },
        Object {
          "name": "HR_Group",
          "type": "property",
          "value": "0",
        },
        Object {
          "name": "HR_Desc",
          "type": "property",
          "value": "",
        },
        Object {
          "name": "HR_State_1",
          "type": "property",
          "value": "",
        },
        Object {
          "name": "HR_State_0",
          "type": "property",
          "value": "",
        },
      ],
      "type": "node",
    },
    Object {
      "given": Array [
        "REGADOR",
        "CHUVA",
      ],
      "node": "GRAMA_MOLHADA",
      "properties": Array [
        Object {
          "name": "data",
          "type": "property",
          "value": Array [
            Array [
              Array [
                0.99,
                0.01,
              ],
              Array [
                0.9,
                0.1,
              ],
            ],
            Array [
              Array [
                0.8,
                0.2,
              ],
              Array [
                0,
                1,
              ],
            ],
          ],
        },
      ],
      "type": "potential",
    },
    Object {
      "given": Array [
        "CHUVA",
      ],
      "node": "REGADOR",
      "properties": Array [
        Object {
          "name": "data",
          "type": "property",
          "value": Array [
            Array [
              0.01,
              0.99,
            ],
            Array [
              0.4,
              0.6,
            ],
          ],
        },
      ],
      "type": "potential",
    },
    Object {
      "given": Array [],
      "node": "CHUVA",
      "properties": Array [
        Object {
          "name": "data",
          "type": "property",
          "value": Array [
            0.2,
            0.8,
          ],
        },
      ],
      "type": "potential",
    },
  ],
  "type": "root",
}
`;
